### ISSUE TITLE
Procedural macro for Vertex trait

### ISSUE THREAD

**Issue by quebin31** [glium/glium#1791]
Is there any reason because Vertex trait cannot be implemented using a procedural macro? Resulting code would be more idiomatic:

**Preview:**
```rust
#[derive(Copy, Clone, Vertex)]
struct Vertex3 {
    position: [f32; 3],
    tex_coords: [f32; 2],
}
```

**Comment by est31** [comment]
Good point! In fact, proc macros would be an awesome idea. Historically, glium used to have a plugin that used to provide such functionality, see the macro directory in the source code (at this point, it should be removed I guess). Also, it's mentioned in the limitations section of the [README](https://github.com/glium/glium/blob/master/README.md):

> Rust plugins are not stable, so you will have to use macros such as `implement_vertex!(MyStruct)` instead of `#[derive(GliumVertex)]`.

Note that both are very old and predate the stabilization of derive macros.

I've wondered about possibly creating a derive macro myself but discarded the idea because glium has a large compile time and compiling a derive macro as well would increase it even further. However, ultimately it's up to the users so we could create a glium_derive crate that users could opt into.

**Comment by quebin31** [comment]
Yeah, creating another crate `glium_derive` and just using a feature would do the trick.

```toml
[dependencies]
glium = { version = "*", feature = ["derive_macros"] }
```

**Comment by est31** [comment]
PRs etc. to do this are definitely welcome. Note that the macros crate has multiple macros so maybe they could serve as inspiration.

**Comment by quebin31** [comment]
`macros` crate inside `glium` isn't being used right? Changes proposed here should be done there?

**Comment by est31** [comment]
It's not used. But I think it's easier to start over. There seems to be a derive impl at https://github.com/voidxnull/glium_derive but it's abandoned. Maybe we can ask @voidxnull to transfer the repo to the glium org.

**Comment by voidxnull** [comment]
@est31 I'll be happy to transfer the repo since I don't use glium anymore.

**Comment by est31** [comment]
@tomaka  would it be possible to facilitate such a transfer?

**Comment by tomaka** [comment]
What should I do? You don't need the approval of the destination to transfer a repo.

**Comment by voidxnull** [comment]
@tomaka This is what I get when I try to transfer the repo: `You don’t have the permission to create public repositories on glium`

**Comment by voidxnull** [comment]
Still can't transfer, it fails with the same error.

**Comment by tomaka** [comment]
It was disabled in the settings! It should work now.

**Comment by voidxnull** [comment]
Done. Now for the maintenance question: I can accept PRs and implement quick fixes, but I'm not sure I'll have much time to implement new features. In any way, I've set up notification properly this time so I'll try to respond as quick as I can.

**Comment by est31** [comment]
@voidxnull can I get write access to the repo? I can help with bringing it up to date, etc.

**Comment by voidxnull** [comment]
@est31 I'm not sure I have the right to do that anymore. The repo settings disappeared for me.

**Comment by Boscop** [comment]
@est31 Any update on this? :)

**Comment by est31** [comment]
@Boscop see https://github.com/glium/glium_derive/issues/3

**Comment by Boscop** [comment]
@est31 Do you have write-access to the `glium_derive` repo now?

**Comment by est31** [comment]
It seems so yeah. No crates.io access but maybe @voidxnull can give me access. Feel free to send PRs in the meantime.

**Comment by voidxnull** [comment]
@Boscop I've managed to fix the warning. Published as 0.1.4.
@est31 Access granted.
